### ISSUE TITLE
[Bug]: Reading AlignedDynamicTable from disk causes hdf5 error

### ISSUE THREAD

**Issue by rly** [hdmf-dev/hdmf#918]
### What happened?

This line https://github.com/hdmf-dev/hdmf/blob/dev/src/hdmf/common/alignedtable.py#L72
in `AlignedDynamicTable.__init__` appends ids to the empty `self.id` if there are custom categories and the user has not set ids. However, if `self.id.data` is an `h5py.Dataset`, this will fail, because more often than not, that column is not chunked and resizable. 

The `intracellular_recordings` table appears to be empty:
<img width="236" alt="image" src="https://github.com/hdmf-dev/hdmf/assets/310197/2cfe3d65-b0a3-4f47-896f-5b979710e9ed">

Data file: 
[test_new_icephys.zip](https://github.com/hdmf-dev/hdmf/files/12165382/test_new_icephys.zip)

### Steps to Reproduce

```shell
import pynwb
io = pynwb.NWBHDF5IO("/Users/rly/Downloads/test_new_icephys.nwb", "r", load_namespaces=True)
nwbfile = io.read()
```

### Traceback

```pytb
Traceback (most recent call last):
  File "/Users/rly/Documents/NWB/hdmf/src/hdmf/build/objectmapper.py", line 1256, in construct
    obj = self.__new_container__(cls, builder.source, parent, builder.attributes.get(self.__spec.id_key()),
  File "/Users/rly/Documents/NWB/hdmf/src/hdmf/build/objectmapper.py", line 1269, in __new_container__
    obj.__init__(**kwargs)
  File "/Users/rly/Documents/NWB/hdmf/src/hdmf/utils.py", line 644, in func_call
    return func(args[0], **pargs)
  File "/Users/rly/Documents/NWB/pynwb/src/pynwb/icephys.py", line 507, in __init__
    super().__init__(**kwargs)
  File "/Users/rly/Documents/NWB/hdmf/src/hdmf/utils.py", line 644, in func_call
    return func(args[0], **pargs)
  File "/Users/rly/Documents/NWB/hdmf/src/hdmf/common/alignedtable.py", line 61, in __init__
    super().__init__(**kwargs)
  File "/Users/rly/Documents/NWB/hdmf/src/hdmf/utils.py", line 644, in func_call
    return func(args[0], **pargs)
  File "/Users/rly/Documents/NWB/hdmf/src/hdmf/common/table.py", line 397, in __init__
    raise ValueError("Must supply 'columns' if specifying 'colnames'")
ValueError: Must supply 'columns' if specifying 'colnames'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/rly/Documents/NWB/hdmf/src/hdmf/utils.py", line 644, in func_call
    return func(args[0], **pargs)
  File "/Users/rly/Documents/NWB/pynwb/src/pynwb/__init__.py", line 304, in read
    file = super().read(**kwargs)
  File "/Users/rly/Documents/NWB/hdmf/src/hdmf/backends/hdf5/h5tools.py", line 477, in read
    return super().read(**kwargs)
  File "/Users/rly/Documents/NWB/hdmf/src/hdmf/utils.py", line 644, in func_call
    return func(args[0], **pargs)
  File "/Users/rly/Documents/NWB/hdmf/src/hdmf/backends/io.py", line 60, in read
    container = self.__manager.construct(f_builder)
  File "/Users/rly/Documents/NWB/hdmf/src/hdmf/utils.py", line 644, in func_call
    return func(args[0], **pargs)
  File "/Users/rly/Documents/NWB/hdmf/src/hdmf/build/manager.py", line 284, in construct
    result = self.__type_map.construct(builder, self, None)
  File "/Users/rly/Documents/NWB/hdmf/src/hdmf/utils.py", line 644, in func_call
    return func(args[0], **pargs)
  File "/Users/rly/Documents/NWB/hdmf/src/hdmf/build/manager.py", line 795, in construct
    return obj_mapper.construct(builder, build_manager, parent)
  File "/Users/rly/Documents/NWB/hdmf/src/hdmf/utils.py", line 644, in func_call
    return func(args[0], **pargs)
  File "/Users/rly/Documents/NWB/hdmf/src/hdmf/build/objectmapper.py", line 1226, in construct
    subspecs = self.__get_subspec_values(builder, self.spec, manager)
  File "/Users/rly/Documents/NWB/hdmf/src/hdmf/build/objectmapper.py", line 1155, in __get_subspec_values
    self.__get_sub_builders(groups, spec.groups, manager, ret)
  File "/Users/rly/Documents/NWB/hdmf/src/hdmf/build/objectmapper.py", line 1206, in __get_sub_builders
    ret.update(self.__get_subspec_values(sub_builder, subspec, manager))
  File "/Users/rly/Documents/NWB/hdmf/src/hdmf/build/objectmapper.py", line 1155, in __get_subspec_values
    self.__get_sub_builders(groups, spec.groups, manager, ret)
  File "/Users/rly/Documents/NWB/hdmf/src/hdmf/build/objectmapper.py", line 1206, in __get_sub_builders
    ret.update(self.__get_subspec_values(sub_builder, subspec, manager))
  File "/Users/rly/Documents/NWB/hdmf/src/hdmf/build/objectmapper.py", line 1155, in __get_subspec_values
    self.__get_sub_builders(groups, spec.groups, manager, ret)
  File "/Users/rly/Documents/NWB/hdmf/src/hdmf/build/objectmapper.py", line 1208, in __get_sub_builders
    ret[subspec] = manager.construct(sub_builder)
  File "/Users/rly/Documents/NWB/hdmf/src/hdmf/utils.py", line 644, in func_call
    return func(args[0], **pargs)
  File "/Users/rly/Documents/NWB/hdmf/src/hdmf/build/manager.py", line 280, in construct
    result = self.__type_map.construct(builder, self, parent)
  File "/Users/rly/Documents/NWB/hdmf/src/hdmf/utils.py", line 644, in func_call
    return func(args[0], **pargs)
  File "/Users/rly/Documents/NWB/hdmf/src/hdmf/build/manager.py", line 795, in construct
    return obj_mapper.construct(builder, build_manager, parent)
  File "/Users/rly/Documents/NWB/hdmf/src/hdmf/utils.py", line 644, in func_call
    return func(args[0], **pargs)
  File "/Users/rly/Documents/NWB/hdmf/src/hdmf/build/objectmapper.py", line 1260, in construct
    raise ConstructError(builder, msg) from ex
hdmf.build.errors.ConstructError: (root/general/intracellular_ephys/intracellular_recordings GroupBuilder {'attributes': {'categories': array(['electrodes', 'stimiuli', 'responses', 'recording_lab_data'],
      dtype=object), 'colnames': array(['recordings_tag'], dtype=object), 'description': 'A table to group together a stimulus and response from a single electrode and a single simultaneous recording and for storing metadata about the intracellular recording.', 'namespace': 'core', 'neurodata_type': 'IntracellularRecordingsTable', 'object_id': '89d4ed78-c007-434a-a15d-2dc73e5a3621'}, 'groups': {'electrodes': root/general/intracellular_ephys/intracellular_recordings/electrodes GroupBuilder {'attributes': {'colnames': array(['electrode', 'voltage_threshold'], dtype=object), 'description': 'Table for storing intracellular electrode related metadata.', 'namespace': 'core', 'neurodata_type': 'IntracellularElectrodesTable', 'object_id': 'b4437b56-857c-45b5-a13e-ee0394a25414'}, 'groups': {}, 'datasets': {'electrode': root/general/intracellular_ephys/intracellular_recordings/electrodes/electrode DatasetBuilder {'attributes': {'description': 'Column for storing the reference to the intracellular electrode', 'namespace': 'hdmf-common', 'neurodata_type': 'VectorData', 'object_id': '1ec83a22-b1a3-41f7-b4e0-f4035bbf38fa'}, 'data': <hdmf.backends.hdf5.h5_utils.BuilderH5ReferenceDataset object at 0x138e4d6f0>}, 'id': root/general/intracellular_ephys/intracellular_recordings/electrodes/id DatasetBuilder {'attributes': {'namespace': 'hdmf-common', 'neurodata_type': 'ElementIdentifiers', 'object_id': 'd582dbc7-8d0a-4afb-9a6f-4e085a7a6e50'}, 'data': <HDF5 dataset "id": shape (3,), type "<i8">}, 'voltage_threshold': root/general/intracellular_ephys/intracellular_recordings/electrodes/voltage_threshold DatasetBuilder {'attributes': {'description': 'Just an example column on the electrodes category table', 'namespace': 'hdmf-common', 'neurodata_type': 'VectorData', 'object_id': 'f84aa5f9-0065-4144-9635-323849eb879d'}, 'data': <HDF5 dataset "voltage_threshold": shape (3,), type "<f8">}}, 'links': {}}, 'recording_lab_data': root/general/intracellular_ephys/intracellular_recordings/recording_lab_data GroupBuilder {'attributes': {'colnames': array(['location'], dtype=object), 'description': 'category table for lab-specific recording metadata', 'namespace': 'hdmf-common', 'neurodata_type': 'DynamicTable', 'object_id': 'fb984967-4d25-41d2-a52f-bc8effc51865'}, 'groups': {}, 'datasets': {'id': root/general/intracellular_ephys/intracellular_recordings/recording_lab_data/id DatasetBuilder {'attributes': {'namespace': 'hdmf-common', 'neurodata_type': 'ElementIdentifiers', 'object_id': '384ca0a3-4504-46f7-af60-d8d0bc846929'}, 'data': <HDF5 dataset "id": shape (3,), type "<i8">}, 'location': root/general/intracellular_ephys/intracellular_recordings/recording_lab_data/location DatasetBuilder {'attributes': {'description': 'Recording location in Middle Earth', 'namespace': 'hdmf-common', 'neurodata_type': 'VectorData', 'object_id': 'aac39e72-4abb-4db7-b912-c8761f8e98e6'}, 'data': <StrDataset for HDF5 dataset "location": shape (3,), type "|O">}}, 'links': {}}, 'responses': root/general/intracellular_ephys/intracellular_recordings/responses GroupBuilder {'attributes': {'colnames': array(['response'], dtype=object), 'description': 'Table for storing intracellular response related metadata.', 'namespace': 'core', 'neurodata_type': 'IntracellularResponsesTable', 'object_id': '72e68c14-f29e-4230-aa4e-fb6a0ccf0cbd'}, 'groups': {}, 'datasets': {'id': root/general/intracellular_ephys/intracellular_recordings/responses/id DatasetBuilder {'attributes': {'namespace': 'hdmf-common', 'neurodata_type': 'ElementIdentifiers', 'object_id': '00ea7d44-d286-48e2-b3e7-d6d6593ad9ec'}, 'data': <HDF5 dataset "id": shape (3,), type "<i8">}, 'response': root/general/intracellular_ephys/intracellular_recordings/responses/response DatasetBuilder {'attributes': {'description': 'Column storing the reference to the recorded response for the recording (rows)', 'namespace': 'core', 'neurodata_type': 'TimeSeriesReferenceVectorData', 'object_id': '3caaefd7-610f-4470-855c-1816b3bb1324'}, 'data': <hdmf.backends.hdf5.h5_utils.BuilderH5TableDataset object at 0x138e40d90>}}, 'links': {}}, 'stimuli': root/general/intracellular_ephys/intracellular_recordings/stimuli GroupBuilder {'attributes': {'colnames': array(['stimulus'], dtype=object), 'description': 'Table for storing intracellular stimulus related metadata.', 'namespace': 'core', 'neurodata_type': 'IntracellularStimuliTable', 'object_id': 'b64409dd-69f2-4b25-926b-5af3724b5502'}, 'groups': {}, 'datasets': {'id': root/general/intracellular_ephys/intracellular_recordings/stimuli/id DatasetBuilder {'attributes': {'namespace': 'hdmf-common', 'neurodata_type': 'ElementIdentifiers', 'object_id': 'd77632c9-f143-4ec7-82a5-ec5098fe5f27'}, 'data': <HDF5 dataset "id": shape (3,), type "<i8">}, 'stimulus': root/general/intracellular_ephys/intracellular_recordings/stimuli/stimulus DatasetBuilder {'attributes': {'description': 'Column storing the reference to the recorded stimulus for the recording (rows)', 'namespace': 'core', 'neurodata_type': 'TimeSeriesReferenceVectorData', 'object_id': 'ed445cdf-f322-4ad4-9681-d4512dc1dc14'}, 'data': <hdmf.backends.hdf5.h5_utils.BuilderH5TableDataset object at 0x138e40790>}}, 'links': {}}}, 'datasets': {'id': root/general/intracellular_ephys/intracellular_recordings/id DatasetBuilder {'attributes': {'namespace': 'hdmf-common', 'neurodata_type': 'ElementIdentifiers', 'object_id': '584735af-e968-468f-8ce5-6ae9a86fe13b'}, 'data': <HDF5 dataset "id": shape (0,), type "<i8">}}, 'links': {}}, "Could not construct IntracellularRecordingsTable object due to: Must supply 'columns' if specifying 'colnames'")
```

### Operating System

macOS

### Python Executable

Conda

### Python Version

3.11

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/hdmf-dev/hdmf/blob/dev/.github/CODE_OF_CONDUCT.md)
- [X] Have you checked the [Contributing](https://github.com/hdmf-dev/hdmf/blob/dev/docs/CONTRIBUTING.rst) document?
- [X] Have you ensured this bug was not already [reported](https://github.com/hdmf-dev/hdmf/issues)?

**Comment by oruebel** [comment]
Normally, this should not happen on `io.read`. However, looking at the file, the issue is triggered here because the file is invalid. There are several errors in this file:

1.  `DynamicTable` and `AlignedDynamicTable` requires that all columns have the same lengths,  however, here the `id` column of the `intracellular_recordings` table is empty but the category tables are not. This triggers error checks in several places where we enforce that columns of tables have the same length.
2. The ``intracellular_recordings.colnames`` specifies  ["recordings_tag"] as a column, but the `VectorData` object for that column is missing. 
3. The `category = ["electrodes","stimiuli","responses","recording_lab_data"]` dataset contains a spelling error where the ``stimuli`` column is misspelled as ``stimiuli``

I talked with @lawrence-mbf  earlier today and it sounds like this may be an error in the icephys tutorial for MatNWB, i.e., I think this is an issue we should fix there. 

If we want to, we could also relax some of the logic in PyNWB to allow us opening bad data files with some of these errors.

For 1., we could modify the `AlignedDynamicTable.__init_` code to relax the code to warn if resizing fails. 

For 2., would require logic in `DynamicTable` to allow reading of files where colnames and the actual columns don't match.

For 3., we would need to autocorrect the `category`, e.g., by removing invalid entries and adding missing entries, which would affect the ordering of categories but should allow opening files with bad `category` data.

**Comment by oruebel** [comment]
The changes in #919 allow reading the file, even though it is invalid. However, it will result in the following set of warnings, since the file is broken. While the file can be opened, it may still result in errors downstream (e.g., when slicing into the table or converting to dataframe) since the columns don't align.


```
src/hdmf/common/table.py:403: UserWarning: Ignoring custom named columns. Must supply 'columns' if specifying 'colnames'=['recordings_tag'] for table name=intracellular_recordings
  warn("Ignoring custom named columns. Must supply 'columns' if specifying"
src/hdmf/common/alignedtable.py:92: RuntimeWarning: Can't resize `id` column for <class 'pynwb.icephys.IntracellularRecordingsTable'>  name=intracellular_recordings due to the following TypeError: Only chunked datasets can be resized
  warnings.warn("Can't resize `id` column for %s  name=%s due to the "
src/hdmf/common/alignedtable.py:103: RuntimeWarning: DynamicTable stimuli does not appear in categories ['electrodes' 'stimiuli' 'responses' 'recording_lab_data']. Using default order ['electrodes', 'recording_lab_data', 'responses', 'stimuli'] categories instead.
  warnings.warn("DynamicTable %s does not appear in categories %s. "
src/hdmf/common/alignedtable.py:129: RuntimeWarning: Category DynamicTable electrodes does not align, it has 3 rows expected 0
  warnings.warn(msg, RuntimeWarning)
src/hdmf/common/alignedtable.py:129: RuntimeWarning: Category DynamicTable recording_lab_data does not align, it has 3 rows expected 0
  warnings.warn(msg, RuntimeWarning)
src/hdmf/common/alignedtable.py:129: RuntimeWarning: Category DynamicTable responses does not align, it has 3 rows expected 0
  warnings.warn(msg, RuntimeWarning)
src/hdmf/common/alignedtable.py:129: RuntimeWarning: Category DynamicTable stimuli does not align, it has 3 rows expected 0
  warnings.warn(msg, RuntimeWarning)
```

**Comment by lawrence-mbf** [comment]
>     1. `DynamicTable` and `AlignedDynamicTable` requires that all columns have the same lengths,  however, here the `id` column of the `intracellular_recordings` table is empty but the category tables are not. This triggers error checks in several places where we enforce that columns of tables have the same length.
> 
>     2. The `intracellular_recordings.colnames` specifies  ["recordings_tag"] as a column, but the `VectorData` object for that column is missing.
> 
>     3. The `category = ["electrodes","stimiuli","responses","recording_lab_data"]` dataset contains a spelling error where the `stimuli` column is misspelled as `stimiuli`
> 

FYI these are resolved as of https://github.com/NeurodataWithoutBorders/matnwb/commit/6cb4d2d01f112b99fdd9477cf5a08deaa47b13ec

The nwbinspector warnings are still present but can be improved later.

**Comment by oruebel** [comment]
> FYI these are resolved as of [NeurodataWithoutBorders/matnwb@6cb4d2d](https://github.com/NeurodataWithoutBorders/matnwb/commit/6cb4d2d01f112b99fdd9477cf5a08deaa47b13ec)

Thanks @lawrence-mbf!
